### ISSUE TITLE
Spurious wakeups with timerfd + Interval

### ISSUE THREAD

**Issue by jonhoo** [DoumanAsh/async-timer#9]
I'm seeing a suspicious number of calls to `poll` when using the `timerfd` implementation combined with `Interval`. In particular, try running
```console
$ cargo t --features tokio_on,stream foobar -- --nocapture
```
With the following test:
```rust
#[cfg(feature = "stream")]
#[tokio::test(threaded_scheduler)]
async fn foobar() {
    use futures_util::stream::StreamExt;

    let mut interval = X(Interval::platform_new(time::Duration::from_millis(2)));

    while let Some(_) = interval.next().await {
        eprintln!("{:?} yield", time::SystemTime::now());
    }
}

use core::future::Future;
use core::{task};
use core::pin::Pin;
use async_timer::oneshot::Oneshot;
struct X<T: Oneshot>(Interval<T>);

#[cfg(feature = "stream")]
impl<T: Oneshot> futures_core::stream::Stream for X<T> {
    type Item = ();

    #[inline]
    fn poll_next(mut self: Pin<&mut Self>, ctx: &mut task::Context) -> task::Poll<Option<Self::Item>> {
        eprintln!("{:?} poll_next", time::SystemTime::now());
        futures_core::Stream::poll_next(Pin::new(&mut self.0), ctx)
    }
}
```
You'll end up with output like the following:
```
SystemTime { tv_sec: 1579742737, tv_nsec: 400813292 } yield
SystemTime { tv_sec: 1579742737, tv_nsec: 400838213 } poll_next
SystemTime { tv_sec: 1579742737, tv_nsec: 400874684 } poll_next
SystemTime { tv_sec: 1579742737, tv_nsec: 400905277 } poll_next
SystemTime { tv_sec: 1579742737, tv_nsec: 402863063 } poll_next
SystemTime { tv_sec: 1579742737, tv_nsec: 402892521 } yield
```
Okay, so a value is yielded at `400813292`. Before it yields, the interval restarts the timer. The future is polled again shortly thereafter (~25µs, at `400838213`) following the call to `next().await`. That call polls the timer, finds that it has not expired, and returns `Pending`. Presumably the next time `poll_next` should be woken up is ~2ms later (as the `Duration` passed to `Interval::new` suggests) when the timer expires. Instead, the future is polled again at `400874684`, ~40µs later, and returns `Pending` again. This happens again at `400905277`, another ~30µs later. Only after that third `poll_next` does the timer appear to "kick in", and the next wakeup to call `poll_next` does not occur until `402863063` (a delay of ~2ms). At that point the interval stream yields `Some` and the patterns starts over again.

These spurious wakeups can be quite costly when an `Interval` is wrapped in a larger `Future`, since that larger `Future`'s `poll` method is what is called, and it might also have to check on a bunch of other sub-`Future`s, since it doesn't know which one was notified.

I'm not sure what is causing this. It does _not_ appear when using the POSIX timer (i.e., if you leave out `tokio_on` from the `--features`), so it's got to have something to do with `timerfd` specifically.

**Comment by DoumanAsh** [comment]
Ok, so I know that tokio tends to do some spurious polls, but it should not be so often...
After wake up, you restart timer by updating to new timestamp and poll it...
Obviously mio would return that it is not ready, otherwise we would clear ready flag https://github.com/DoumanAsh/async-timer/blob/master/src/oneshot/timer_fd.rs#L165

Logically speaking I think I did everything I could, but in case of kevent and timerfd timers I depend on tokio to wake timer.
POSIX timer is using my own waker to wake once (by calling callback from signal)

Out of curiosity could you try the same with single threaded runtime?
I'll try to check tokio integration myself once I'll have time

**Comment by DoumanAsh** [comment]
Note to self: checked macOS, has the same problem.
Type of scheduler doesn't matter

**Comment by jonhoo** [comment]
Hmm, I do remember having some odd issues with using `PollEvented` back in the day, and then ended up [moving to `Registration`](https://github.com/jonhoo/tokio-os-timer/commit/5a840b988cc3be223dfa4b444badc821d93d62ff). I wonder if that might work here too?

**Comment by DoumanAsh** [comment]
yeah I'll check what PollEvented does, but there is definitely no read notification and yet tokio calls in the future.
So probably PollEvented registers more than read interest with mio

**Comment by DoumanAsh** [comment]
Conclusion: `PollEvented` might register non-read interest so we should either not use it or work it around by removing non-read interests

**Comment by DoumanAsh** [comment]
Wait for https://github.com/tokio-rs/tokio/issues/2222

**Comment by DoumanAsh** [comment]
Fixed by tokio and will be available in 0.2.12

**Comment by jonhoo** [comment]
I can confirm that tokio-rs/tokio#2222 indeed fixed this issue!
